python/cuspatial/cuspatial/core/binpreds/feature_touches.py

```diff
@@ -1,50 +1,123 @@
-    """Base class for binary predicates that use the contains predicate
-    to implement the touches predicate. For example, a Point-Polygon
-    Touches predicate is defined in terms of a Point-Polygon Contains
-    predicate.
+    """
+    If any point is shared between the following geometry types, they touch:
 
     Used by:
-    (Point, Polygon)
-    (Polygon, Point)
+    (Point, MultiPoint)
+    (Point, LineString)
+    (MultiPoint, Point)
+    (MultiPoint, MultiPoint)
+    (MultiPoint, LineString)
+    (MultiPoint, Polygon)
+    (LineString, Point)
+    (LineString, MultiPoint)
     (Polygon, MultiPoint)
-    (Polygon, LineString)
-    (Polygon, Polygon)
     """
 
-    pass
+    def _preprocess(self, lhs, rhs):
+        equals = _basic_equals(lhs, rhs)
+        return equals
+
+
+class PointPolygonTouches(ContainsPredicate):
+    def _preprocess(self, lhs, rhs):
+        # Reverse argument order.
+        equals_all = _basic_equals_all(rhs, lhs)
+        touches = _basic_intersects(rhs, lhs)
+        return ~equals_all & touches
+
+
+class LineStringLineStringTouches(BinPred):
+    def _preprocess(self, lhs, rhs):
+        """A and B have at least one point in common, and the common points
+        lie in at least one boundary"""
+        # Point is equal
+        equals = _basic_equals(lhs, rhs)
+        # Linestrings are not equal
+        equals_all = _basic_equals_all(lhs, rhs)
+        # Linestrings do not cross
+        crosses = lhs.crosses(rhs)
+        return equals & ~crosses & ~equals_all
+
+
+class LineStringPolygonTouches(BinPred):
+    def _preprocess(self, lhs, rhs):
+        pli = _basic_intersects_pli(lhs, rhs)
+        if len(pli[1]) == 0:
+            return _false_series(len(lhs))
+        intersections = _points_and_lines_to_multipoints(pli[1], pli[0])
+        # A touch can only occur if the point in the intersection
+        # is equal to a point in the linestring: it must
+        # terminate in the boundary of the polygon.
+        equals = _basic_equals_count(intersections, lhs) > 0
+        intersects = _basic_intersects_count(lhs, rhs)
+        intersects = (intersects == 1) | (intersects == 2)
+        contains = rhs.contains(lhs)
+        contains_any = _basic_contains_properly_any(rhs, lhs)
+        return equals & intersects & ~contains & ~contains_any
+
+
+class PolygonPointTouches(BinPred):
+    def _preprocess(self, lhs, rhs):
+        intersects = _basic_intersects(lhs, rhs)
+        return intersects
+
+
+class PolygonLineStringTouches(LineStringPolygonTouches):
+    def _preprocess(self, lhs, rhs):
+        return super()._preprocess(rhs, lhs)
+
+
+class PolygonPolygonTouches(BinPred):
+    def _preprocess(self, lhs, rhs):
+        contains_lhs_none = _basic_contains_count(lhs, rhs) == 0
+        contains_rhs_none = _basic_contains_count(rhs, lhs) == 0
+        intersects = _basic_intersects_count(lhs, rhs) == 1
+        return contains_lhs_none & contains_rhs_none & intersects
```

Comment:
The following example has more than 1 point in the intersection result. (It's a segment, and `_basic_intersects_count` returns 2 in this case?
```python
p0 = Polygon([(0, 0), (1, 0), (1, 1), (0, 0)])
p1 = Polygon([(0.25, 0.25), (0.5, 0.5), (0, 0.5), (0.25, 0.25)])
p0.touches(p1) # True
```